otter/test/test_cloud_client.py

```diff
@@ -255,6 +272,199 @@ def test_params(self):
         pure_request_eff = resolve_authenticate(eff)
         self.assertEqual(pure_request_eff.intent.params, {"foo": ["bar"]})
 
+    def test_throttling(self):
+        """
+        When the throttler function returns a bracketing function, it's used to
+        throttle the request.
+        """
+        def throttler(stype, method):
+            if stype == ServiceType.CLOUD_SERVERS and method == 'get':
+                return bracket
+        bracket = object()
+        svcreq = service_request(
+            ServiceType.CLOUD_SERVERS, 'GET', 'servers').intent
+
+        response = stub_pure_response({}, 200)
+        seq = SequenceDispatcher([
+            unwrap_wrapped_effect(_Throttle, dict(bracket=bracket), [
+                (Authenticate(authenticator=self.authenticator,
+                              tenant_id=1,
+                              log=self.log),
+                 lambda i: ('token', fake_service_catalog)),
+                (Request(method='GET', url='http://dfw.openstack/servers',
+                         headers=headers('token'), log=self.log),
+                 lambda i: response),
+            ])])
+
+        eff = self._concrete(svcreq, throttler=throttler)
+        with seq.consume():
+            result = sync_perform(seq, eff)
+        self.assertEqual(result, (response[0], {}))
+
+
+class ThrottleTests(SynchronousTestCase):
+    """Tests for :obj:`_Throttle` and :func:`_perform_throttle`."""
+
+    def test_perform_throttle(self):
+        """
+        The bracket given to :obj:`_Throttle` is used to call the nested
+        performer.
+        """
+        def bracket(f, *args, **kwargs):
+            return f(*args, **kwargs).addCallback(lambda r: ('bracketed', r))
+        throttle = _Throttle(bracket=bracket, effect=Effect(Constant('foo')))
+        dispatcher = ComposedDispatcher([
+            TypeDispatcher({_Throttle: _perform_throttle}),
+            base_dispatcher])
+        result = sync_perform(dispatcher, Effect(throttle))
+        self.assertEqual(result, ('bracketed', 'foo'))
+
+
+class SerializeAndDelayTests(SynchronousTestCase):
+    """Tests for :func:`_serialize_and_delay`."""
+
+    @mock.patch('otter.cloud_client.DeferredLock')
+    def test_serialize_and_delay(self, deferred_lock):
+        """
+        :func:`_serialize_and_delay` returns a function that, when given a
+        function and arguments, calls it inside of a lock and after a specified
+        delay.
+        """
+        class DeferredLock(object):
+            def run(self, f, *args, **kwargs):
+                return f(*args, **kwargs).addCallback(lambda r: ('locked', r))
+        deferred_lock.side_effect = DeferredLock
```

Comment:
Non-blocking: Do you think that test without mocking `DeferredLock` that calls `serialize_and_delay` twice and notices that they get serialized and delayed would be better? I can understand if you don't want to spend much time on the tests since this implementation should be going away soon :).